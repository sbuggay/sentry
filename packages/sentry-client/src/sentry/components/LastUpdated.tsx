import * as React from "react";
import { connect } from "react-redux";

interface ILastUpdatedProps {
    lastUpdated?: number
}

export class LastUpdated extends React.Component<any, any> {

    getStyle() {
        return {
            fontSize: "14px",
            color: "#adb0af"
        }
    }

    render(): JSX.Element {
        return (
            <div style={this.getStyle()}>
                Last updated {new Date(this.props.lastUpdated).toTimeString()}
            </div>
        );
    }
}

const mapStateToProps = (state: any) => {
    return {
        lastUpdated: state.lastUpdated
    };
};

export default connect(mapStateToProps)(LastUpdated);
