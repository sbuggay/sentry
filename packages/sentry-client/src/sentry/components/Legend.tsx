import * as React from "react";

import { Status } from "./Status";
import { STATUS } from "../constants/status";

export class Legend extends React.Component<any, any> {

    public getStyle() {
        return {
            fontSize: "14px",
            color: "#adb0af"
        };
    }

    public render(): JSX.Element {
        // TODO: Clean up this stupid style hack
        return (
            <div style={this.getStyle()}>
                <span><Status status={STATUS.AVAILABLE} /> Available</span>
                <span style={{marginLeft: "1em"}}><Status status={STATUS.OUTAGE} /> Outage</span>
                <span style={{marginLeft: "1em"}}><Status status={STATUS.ISSUE} /> Issue</span>
                <span style={{marginLeft: "1em"}}><Status status={STATUS.MAINTENANCE} /> Maintenance</span>
                <span style={{marginLeft: "1em"}}><Status status={STATUS.UNKNOWN} /> Unknown</span>
            </div>
        );
    }
}

export default Legend;
