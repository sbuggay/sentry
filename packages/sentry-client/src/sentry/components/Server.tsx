import * as React from "react";
import { removeServer } from "../actions";
import { connect } from "react-redux";

import { IServer } from "../reducer";

import Status from "./Status";
import { STATUS } from "../constants/status";

import { formatBytes } from "../lib/utils";

interface IStateProps {
    server: IServer;
    index: Number;
    expanded?: Boolean;
};

interface IDispatchProps {
    removeServer?: Function;
};

export class Server extends React.Component<IStateProps & IDispatchProps, any> {
    constructor(props: any) {
        super(props);
        this.state = {
            expanded: false,
            hover: false
        }
    }

    //Handlers
    handleClick() {
        this.toggleExpand();
    }

    handleKeyDown(e: React.KeyboardEvent<HTMLDivElement>) {
        switch (e.key) {
            case "Enter":
                this.toggleExpand();
                break;
        }
    }

    handleOnMouseEnter() {
        this.setState({
            hover: true
        });
    }

    handleOnMouseLeave() {
        this.setState({
            hover: false
        });
    }

    toggleExpand() {
        this.setState({
            expanded: !this.state.expanded
        })
    }

    getStyle(): React.CSSProperties{
        return {
            width: "280px",
            padding: "5px 10px",
            borderBottom: "1px solid #adb0af"
        };
    }

    //Render
    renderData(): JSX.Element {
        const staticInfo = this.props.server.staticInfo;
        const dynamicInfo = this.props.server.dynamicInfo;

        const cpu = dynamicInfo.cpus[0].model;

        const renderRow = (label: string, data: string) => {
            return (
                <div style={{display: "flex", justifyContent: "space-between"}}>
                    <span>{label}</span>
                    <span>{data}</span>
                </div>
            );
        }

        const cpuModel = dynamicInfo.cpus[0].model.split("@")[0];
        const cpuSpeed = dynamicInfo.cpus[0].model.split("@")[1];
        const virtualCores = dynamicInfo.cpus.length;

        return (
            <div style={{marginTop: "0.5em", overflow: "hidden"}}>
                {renderRow("hostname:", dynamicInfo.hostname)}
                {renderRow("uptime:", dynamicInfo.uptime.toString())}
                {renderRow("cpu:", cpuModel)}
                {renderRow("", `${cpuSpeed} (${virtualCores})`)}
                {renderRow("ram:", `${formatBytes(dynamicInfo.freemem)} / ${formatBytes(dynamicInfo.totalmem)}`)}
            </div>
        );
    }

    renderServiceData(): JSX.Element {
        const services = this.props.server.serviceInfo;

        if (!services) {
            return null;
        }

        const serviceKeys = Object.keys(services);

        return (
            <div style={{marginTop: "0.5em"}}>
                <div>services:</div>
                {serviceKeys.map((key, index) => {
                    const status = services[key].status ? STATUS.AVAILABLE : STATUS.OUTAGE;
                    return (
                        <div key={index}>
                            <Status status={status} />
                            {services[key].name}
                        </div>
                    )
                })}
            </div>
        );
    }

    renderDetails(): JSX.Element {

        function detailStyle() {
            return {
                fontSize: "14px",
                marginBottom: "10px"
            }
        }

        return (
            <div style={detailStyle()}>
                {this.renderData()}
                {this.renderServiceData()}
            </div>
        );
    }

    renderChevron(): JSX.Element {

        const style = {
            marginLeft: "10px",
            float: "right"
        };

        const className = this.state.expanded ? "fa fa-chevron-down" : "fa fa-chevron-right";

        return (
            <i 
            onClick={() => this.handleClick()}
            style={style} 
            className={className}></i>
        );
    }

    render(): JSX.Element {
        return (
            <div
                tabIndex={0}
                onMouseEnter={() => this.handleOnMouseEnter()}
                onMouseLeave={() => this.handleOnMouseLeave()}
                onKeyDown={(e) => this.handleKeyDown(e)}
                style={this.getStyle()}>
                <div>
                    <span>
                        <Status status={this.props.server.status} />
                        {this.props.server.name}
                    </span>
                    {this.state.hover ? this.renderChevron() : null}
                </div>
                {this.state.expanded ? this.renderDetails() : null}
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch: Function) => {
    return {
        removeServer: () => dispatch(removeServer)
    };
};

export default connect<IStateProps, IDispatchProps, any>(undefined, mapDispatchToProps)(Server);
