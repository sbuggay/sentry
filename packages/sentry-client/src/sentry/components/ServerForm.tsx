import * as React from "react";

interface IProps {
	handleSubmit?: Function;
	onSubmit?: Function;
};

export default class ServerForm extends React.Component<IProps, any> {
	constructor(props: any) {
		super(props);

		this.state = {
			nameValue: "",
			hostValue: ""
		}
	}

	handleChange(event: any) {
		const target = event.target;
		const value = target.type === 'checkbox' ? target.checked : target.value;
		const name = target.name;
		this.setState({ [name]: target.value });
	}

	handleSubmit(event: any) {
		this.props.handleSubmit(this.state);
		event.preventDefault();
	}

	render() {
		return (
			<form onSubmit={this.handleSubmit.bind(this)}>
				<input placeholder="name" type="text" name="nameValue" value={this.state.nameValue} onChange={this.handleChange.bind(this)}></input>
				<input placeholder="host" type="text" name="hostValue" value={this.state.hostValue} onChange={this.handleChange.bind(this)}></input>
				<button type="submit">submit</button>
			</form>
		);
	}
};
