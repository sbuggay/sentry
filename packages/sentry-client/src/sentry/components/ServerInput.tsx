import * as React from "react";
import { Field, reduxForm } from "redux-form";

interface IStateProps {
	handleSubmit: any;
}

export class ServerInput extends React.Component<IStateProps, any> {
	render() {
		const { handleSubmit } = this.props;
		return (
			<form onSubmit={handleSubmit}>
				<Field name="hostname" component="input" type="text" />
				<button type="submit">Submit</button>
			</form>
		);
	}
}

export default reduxForm({ form: "hostnameInput" })(ServerInput);