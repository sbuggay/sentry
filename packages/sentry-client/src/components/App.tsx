import * as React from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";

import List from "./List";
import LastUpdated from "./LastUpdated";
import Legend from "./Legend";
import Storage from "./Storage";
import GroupSelect from "./GroupSelect";

import { EView } from "../constants";

import { initialize, addServer, changeView } from "../actions";

interface IStateProps {
    title?: string;
    view?: EView;
}

interface IDispatchProps {
    initialize?: () => any;
    addServer?: (values: object) => any;
    changeView?: (view: EView) => void;
}

export class App extends React.Component<IStateProps & IDispatchProps, any> {

    public componentWillMount() {
        if (this.props.initialize) {
            this.props.initialize();
        }
    }

    public onServerInputSubmit(values: object) {
        if (this.props.addServer) {
            this.props.addServer(values);
        }
    }

    public getStyle() {
        return {
            maxWidth: "900px",
            marginLeft: "auto",
            marginRight: "auto"
        };
    }

    public handleSelectChange(index: EView) {
        if (this.props.changeView) {
            this.props.changeView(index);
        }
    }

    public render() {

        const options = [];
        for (const option in EView) {
            options.push(EView[option]);
        }

        return (
            <div>
                <h1 style={{ textAlign: "center", fontWeight: 500 }}>
                    {this.props.title}
                </h1>
                <div style={this.getStyle()}>
                    <div style={{ display: "flex", justifyContent: "space-between" }}>
                        <div></div>
                         <GroupSelect
                            onChange={(index) => this.handleSelectChange(index)}
                            options={options}
                            selectedOption={this.props.view || EView.servers} />
                    </div>
                    <List />
                    <div style={{ display: "flex", justifyContent: "space-between" }}>
                        <LastUpdated />
                        <Legend />
                    </div>
                    <div>
                        <Storage />
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state: any) => {
    return {
        title: state.title,
        view: state.view
    };
};

const mapDispatchToProps = (dispatch: any) => {
    return bindActionCreators({
        addServer,
        initialize,
        changeView
    }, dispatch);
};

export default connect<IStateProps & IDispatchProps, any, any>(mapStateToProps, mapDispatchToProps)(App);
