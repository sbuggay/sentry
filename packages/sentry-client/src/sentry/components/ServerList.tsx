import * as React from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";

import Server from "./Server";
import { initializePolling, pollServers } from "../actions";
import { IServer } from "../reducer";

interface IServerListProps {
    servers?: {
        [id: string]: IServer
    };
    initializePolling?: () => any;
    pollServers?: () => any;
}

// TODO: Figure out what the actual type of a component ref is
interface IServerListState {
    refs: any[];
}

export class ServerList extends React.Component<IServerListProps, IServerListState> {

    public componentWillMount() {
        if (this.props.pollServers) {
            this.props.pollServers();
        }
        if (this.props.initializePolling) {
            this.props.initializePolling();
        }
    }

    public getStyle() {
        return {
            display: "flex",
            flexFlow: "row wrap",
            width: "100%",
            margin: "1.5em 0 1.5em"
        };
    }

    public render() {
        return (
            <div style={this.getStyle()}>
                {Object.keys(this.props.servers).map((id, index) => {
                    if (!this.props.servers || !this.props.servers[id]) {
                        return null;
                    }

                    return <Server
                        index={index}
                        key={index}
                        server={this.props.servers[id]} />;
                })}
            </div>
        );
    }
}

const mapStateToProps = (state: any) => {
    return {
        servers: state.servers
    };
};

const mapDispatchToProps = (dispatch: any) => {
    return bindActionCreators({
        pollServers,
        initializePolling
    }, dispatch);
};

export default connect<IServerListProps, any, any>(mapStateToProps, mapDispatchToProps)(ServerList);
