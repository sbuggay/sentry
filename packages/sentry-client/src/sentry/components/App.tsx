import * as React from "react";
import { connect } from "react-redux";

import ServerList from "./ServerList";
import ServerForm from "./ServerForm";
import Status from "./Status";

import { addServer } from "../actions";

import { guid } from "../lib/utils";

import { STATUS } from "../constants/status";

interface IDispatchProps {
	addServer?: Function;
};

class App extends React.Component<IDispatchProps, any> {
	onServerInputSubmit(values: Object) {
		console.log(values);
		this.props.addServer(values);
	}

	render() {
		return <div>
			<Status status={STATUS.AVAILABLE} />
			<Status status={STATUS.ISSUE} />
			<Status status={STATUS.OUTAGE} />
			<Status status={STATUS.MAINTENANCE} />
			<ServerForm handleSubmit={this.onServerInputSubmit.bind(this)} />
			<ServerList />
		</div>;
	}
}

const mapDispatchToProps = (dispatch: any) => {
	return {
		addServer: (values: any) => {
			dispatch(addServer(values.nameValue, values.hostValue, guid()));
		}
	};
};

export default connect<any, IDispatchProps, any>(undefined, mapDispatchToProps)(App);