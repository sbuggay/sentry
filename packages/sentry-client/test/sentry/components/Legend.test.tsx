import * as React from "react";
import * as renderer from "react-test-renderer";
import { Legend } from "../../../src/sentry/components/Legend";

describe("Legend", () => {
	it("renders a legend", () => {
		const component = renderer.create(
			<Legend />
		);

		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
