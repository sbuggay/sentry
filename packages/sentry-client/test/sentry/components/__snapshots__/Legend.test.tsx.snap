// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Legend renders a legend 1`] = `
<div
  style={
    Object {
      "color": "#adb0af",
      "fontSize": "14px",
    }
  }
>
  <span>
    <i
      aria-hidden="true"
      className="fa fa-circle"
      style={
        Object {
          "color": "#229926",
          "marginRight": "0.5em",
        }
      }
    />
     Available
  </span>
  <span
    style={
      Object {
        "marginLeft": "1em",
      }
    }
  >
    <i
      aria-hidden="true"
      className="fa fa-exclamation-triangle"
      style={
        Object {
          "color": "#951A1D",
          "marginRight": "0.5em",
        }
      }
    />
     Outage
  </span>
  <span
    style={
      Object {
        "marginLeft": "1em",
      }
    }
  >
    <i
      aria-hidden="true"
      className="fa fa-exclamation-circle"
      style={
        Object {
          "color": "#FEE032",
          "marginRight": "0.5em",
        }
      }
    />
     Issue
  </span>
  <span
    style={
      Object {
        "marginLeft": "1em",
      }
    }
  >
    <i
      aria-hidden="true"
      className="fa fa-square"
      style={
        Object {
          "color": "#1172C6",
          "marginRight": "0.5em",
        }
      }
    />
     Maintenance
  </span>
  <span
    style={
      Object {
        "marginLeft": "1em",
      }
    }
  >
    <i
      aria-hidden="true"
      className="fa fa-circle"
      style={
        Object {
          "color": "gray",
          "marginRight": "0.5em",
        }
      }
    />
     Unknown
  </span>
</div>
`;
